packages/framework/src/Framework/Actions/CreatesNewPublicationPage.php

```diff
@@ -0,0 +1,96 @@
+ */
+class CreatesNewPublicationPage extends CreateAction implements CreateActionContract
+{
+    public function __construct(
+        protected PublicationType $pubType,
+        protected Collection $fieldData,
+        protected bool $force = false,
+        protected ?OutputStyle $output = null,
+    ) {
+        $canonicalFieldName = $this->pubType->canonicalField;
+        $canonicalFieldDefinition = $this->pubType->getCanonicalFieldDefinition();
+        $canonicalValue = $this->getCanonicalValue($canonicalFieldDefinition, $canonicalFieldName);
+        $canonicalStr = Str::of($canonicalValue)->substr(0, 64);
+
+        $fileName = $this->formatStringForStorage($canonicalStr->slug()->toString());
+        $directory = $this->pubType->getDirectory();
+        $this->outputPath = "$directory/$fileName.md";
+    }
+
+    protected function handleCreate(): void
+    {
+        $now = Carbon::now()->format('Y-m-d H:i:s');
+        $output = "---\n";
+        $output .= "__createdAt: $now\n";
+        foreach ($this->fieldData as $name => $value) {
+            /** @var PublicationField $fieldDefinition */
+            $fieldDefinition = $this->pubType->getFields()->where('name', $name)->firstOrFail();
+
+            if ($fieldDefinition->type === PublicationFieldTypes::Text) {
+                $output .= "$name: |\n";
+                if (is_string($value)) {
+                    $value = Str::of($value)->explode("\n");
+                }
+                foreach ($value as $line) {
+                    $output .= "  $line\n";
+                }
+                continue;
+            }
+
+            if ($fieldDefinition->type === PublicationFieldTypes::Array) {
+                $output .= "$name:\n";
+                foreach ($value as $item) {
+                    $output .= "  - \"$item\"\n";
+                }
+                continue;
+            }
+
+            $output .= "$name: $value\n";
+        }
+        $output .= "---\n";
+        $output .= "\n## Write something awesome.\n\n";
+
+        $this->output?->writeln("Saving publication data to [$this->outputPath]");
+
+        $this->save($output);
+    }
+
+    protected function getCanonicalValue(PublicationField $canonicalFieldDefinition, string $canonicalFieldName): string
+    {
+        if ($canonicalFieldName === '__createdAt') {
+            return Carbon::now()->format('Y-m-d H:i:s');
+        }
+
+        // TODO: Is it reasonable to use arrays as canonical field values?
+        if ($canonicalFieldDefinition->type === PublicationFieldTypes::Array) {
+            $canonicalValue = ($this->fieldData->get($canonicalFieldName) ?? [])[0];
+        } else {
+            $canonicalValue = $this->fieldData->get($canonicalFieldName);
+        }
+
+        return $canonicalValue ?? throw new RuntimeException("Could not find field value for '$canonicalFieldName' which is required for as it's the type's canonical field", 404);
+    }
```

Comment:
In a recent PR I made so that some types could not be selected as canonical when using the make pubType command, however since files can be edited arbitrary, other types can be set manually. We may need to validate this when parsing schema files.


_Originally posted by @caendesilva in https://github.com/hydephp/develop/issues/774#issuecomment-1364586165_

Reply:
Maybe we want to validate the schema when running the validate command? In either case, I don't think we need a validation just for canonical fields. The worst that can happen is that things won't compile or there will be conflicts.